be/src/exec/workgroup/work_group.cpp

```diff
@@ -89,12 +97,206 @@ void WorkGroupManager::remove_workgroup(int wg_id) {
+        // so we just reduce resource of the workgroup which it over pay by ratio
+        // and add resource of the workgroup which it short of
+        for (auto const& wg : _io_wgs) {
+            if (wg->get_diff_factor() < 0) {
+                wg->update_select_factor(0 - negative_total_diff_factor * wg->get_diff_factor() /
+                                                     negative_total_diff_factor);
+            } else if (wg->get_diff_factor() > 0) {
+                wg->update_select_factor(0 - negative_total_diff_factor * wg->get_diff_factor() /
+                                                     positive_total_diff_factor);
+            }
+        }
+    } else {
+        // if positive_total_diff_factor + negative_total_diff_factor <= 0
+        // it mean available resource less than short of resource, but exist some workgroup which over pay
+        // so we just reduce resource of the workgroup which it over pay by ratio
+        // and add resource of the workgroup which it short of
+        for (auto const& wg : _io_wgs) {
+            if (wg->get_diff_factor() < 0) {
+                wg->update_select_factor(positive_total_diff_factor * wg->get_diff_factor() /
+                                         negative_total_diff_factor);
+            } else if (wg->get_diff_factor() > 0) {
+                wg->update_select_factor(0 - positive_total_diff_factor * wg->get_diff_factor() /
+                                                     positive_total_diff_factor);
+            }
+        }
+    }
+    _cur_schedule_num = _schedule_num_period;
+    _is_schedule.store(false, std::memory_order_release);
+}
+
+WorkGroupPtr WorkGroupManager::get_next_wg() {
+    size_t index = _cur_index;
+    while (!_cur_index.compare_exchange_strong(index, index + 1)) {
+        index = _cur_index;
+        if (index >= _cur_wait_run_wgs.size()) {
+            break;
+        }
+    }
+
+    if (index < _cur_wait_run_wgs.size()) {
+        return _cur_wait_run_wgs[index];
+    }
+
+    bool expect = false;
+    if (!_is_schedule.compare_exchange_strong(expect, true)) {
+        // maybe has other thread do it
+        return nullptr;
+    }
+
+    // during schedule period, we pre calculate next period wg for speed up
+    for (size_t i = 0; i < _schedule_num_period; i++) {
+        size_t idx = get_next_wg_index();
+        _cur_wait_run_wgs[i] = _workgroups[idx];
+    }
+    _cur_index = 1;
+    index = 0;
+    _is_schedule.store(false, std::memory_order_release);
+    return _cur_wait_run_wgs[index];
+}
+
+size_t WorkGroupManager::get_next_wg_index() {
+    // we use Weighted round robin
+    size_t index = -1;
+    double total = 0;
+    for (int i = 0; i < _io_wgs.size(); i++) {
+        _io_wgs[i]->update_cur_select_factor(_io_wgs[i]->get_select_factor());
+        total += _io_wgs[i]->get_select_factor();
+        if (index == -1 || _io_wgs[i]->get_cur_select_factor() < _io_wgs[i]->get_cur_select_factor()) {
```

Comment:
_io_wgs[i]->get_cur_select_factor() < _io_wgs[i]->get_cur_select_factor() is invalid?
the right form is 
```
_io_wgs[i]->get_cur_select_factor() < _io_wgs[index]->get_cur_select_factor() 
```
or 

```
_io_wgs[index]->get_cur_select_factor() < _io_wgs[i]->get_cur_select_factor() 
```